airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py

```diff
@@ -0,0 +1,180 @@
+
+        for dimension in response.dimension_headers:
+            properties[dimension.name] = {"type": "string"}
+
+        for metric in response.metric_headers:
+            properties[metric.name] = {"type": "number"}
+
+        json_schema = {
+            "$schema": "http://json-schema.org/draft-07/schema#",
+            "type": "object",
+            "properties": properties,
+        }
+
+        primary_key = list(map(lambda h: [h.name], response.dimension_headers))
+
+        stream = AirbyteStream(
+            name=report_name,
+            json_schema=json_schema,
+            supported_sync_modes=[SyncMode.full_refresh, SyncMode.incremental],
+            source_defined_primary_key=primary_key,
+            default_cursor_field=[DEFAULT_CURSOR_FIELD]
+        )
+        return AirbyteCatalog(streams=[stream])
+
+    def read(
+        self, logger: logging.Logger, config: Mapping[str, Any], catalog: ConfiguredAirbyteCatalog, state: MutableMapping[str, Any] = None
+    ) -> Generator[AirbyteMessage, None, None]:
+        """
+        Returns a generator of the AirbyteMessages generated by reading the source with the given configuration,
+        catalog, and state.
+
+        :param logger: Logging object to display debug/info/error to the logs
+            (logs will not be accessible via airbyte UI if they are not passed to this logger)
+        :param config: Json object containing the configuration of this source, content of this json is as specified in
+            the properties of the spec.json/spec.yaml file
+        :param catalog: The input catalog is a ConfiguredAirbyteCatalog which is almost the same as AirbyteCatalog
+            returned by discover(), but
+        in addition, it's been configured in the UI! For each particular stream and field, there may have been provided
+        with extra modifications such as: filtering streams and/or columns out, renaming some entities, etc
+        :param state: When a Airbyte reads data from a source, it might need to keep a checkpoint cursor to resume
+            replication in the future from that saved checkpoint.
+            This is the object that is provided with state from previous runs and avoid replicating the entire set of
+            data everytime.
+
+        :return: A generator that produces a stream of AirbyteRecordMessage contained in AirbyteMessage object.
+        """
+        report_name = config.get("report_name")
+
+        response = self._run_report(config)
+
+        dimensions = list(map(lambda h: h.name, response.dimension_headers))
+        metrics = list(map(lambda h: h.name, response.metric_headers))
+
+        last_cursor_value = state.get(report_name, {}).get(DEFAULT_CURSOR_FIELD, "")
+
+        for row in response.rows:
+            data = dict(zip(dimensions, list(map(lambda v: v.value, row.dimension_values))))
+            data.update(dict(zip(metrics, list(map(lambda v: float(v.value), row.metric_values)))))
```

Comment:
we can move these data processes to that client

Reply:
Moved here https://github.com/airbytehq/airbyte/pull/12701/commits/95b2855db98c01c174bf527a3e83fbafb5285f6a.